src/v/cluster/archival/segment_reupload.cc

```diff
@@ -202,6 +203,8 @@ void segment_collector::do_collect(segment_collector_mode mode) {
 
     align_end_offset_to_manifest(
       _target_end_inclusive.value_or(current_segment_end));
+
+    maybe_clamp_end_to_term_boundary();
```

Comment:
I'm reading the code to try to understand the implications of this change and have an idea for an alternative implementation, likely much more cleaner too. What if we don't clamp the boundary here but instead (!) we add a check here https://github.com/redpanda-data/redpanda/blob/a9cfe591f7bc8fb5d2eb9f4a664c23f9a0fde1c8/src/v/cluster/archival/segment_reupload.cc#L193 `if (!_segments.empty() && _segments.back()->offsets().get_term() != result.segment->offsets().get_term()) {_end_inclusive=...; break;}`

Would this be enough? And this is where

Reply:
Don't know what happened there. The message should still make sense without that last part.

Reply:
I think it's more readable in a separate method.

Reply:
these failures are important, if we failed to call `_log.get_term(_begin_inclusive)` then it's not safe to start the upload

Reply:
I agree but this isn't helping much. Even after this code is done we can still have concurrent truncations making all these checks useless.